### PR TITLE
HybridFactorGraph emplace variants

### PR DESCRIPTION
**Pull request by dellaert**
emplace variants, automatically deduce the right graph.

Unfortunately emplace variants don't deal well with {}, but it works with auto.

### CODE REVIEW

**Comment by dellaert** [comment]
Review for @ProfFan , cc @varunagrawal

File: gtsam/hybrid/HybridFactorGraph.cpp [hunk 1]
```diff
@@ -35,10 +35,12 @@ void HybridFactorGraph::print(const std::string &str,
   std::string nonlinearStr = str + ": NonlinearFactorGraph";
   std::string discreteStr = str + ": DiscreteFactorGraph";
   std::string dcStr = str + ": DCFactorGraph";
+  std::string gStr = str + ": GaussianGraph";
```
Comment (ProfFan) [inline gtsam/hybrid/HybridFactorGraph.cpp:38]: const

File: gtsam/hybrid/HybridFactorGraph.h [hunk 2]
```diff
@@ -53,6 +53,57 @@ class HybridFactorGraph {
...
+
+  /// Construct a factor and add (shared pointer to it) to factor graph.
+  template <class FACTOR, class... Args>
+  IsGaussian<FACTOR> emplace_shared(Args&&... args) {
+    gaussianGraph_.push_back(boost::allocate_shared<FACTOR>(
+        Eigen::aligned_allocator<FACTOR>(), std::forward<Args>(args)...));
+  }
+
+  // TODO(dellaert): from below I think we should only keep shared pointer
+  // versions.
```
Comment (ProfFan) [inline gtsam/hybrid/HybridFactorGraph.h:105]: I agree, since I just looked at GFG and NFG and they all use `shared_ptr` internally.

**Review by ProfFan — APPROVED** [review]

**Comment by varunagrawal** [comment]
I'll wait for this PR to merge before making any updates since I need some of the changes here. :slightly_smiling_face:

File: gtsam/hybrid/HybridFactorGraph.cpp (hunk 1 above)
Comment (dellaert) [reply gtsam/hybrid/HybridFactorGraph.cpp:38]: not even sure why we have these intermediate variables. Feel free to move them in place after I merge.

**Comment by dellaert** [comment]
merged before checks finish. I think the arg forwarding might fail on some platforms, but we can handle in the linearize PR

**Pull request merged**